### DISCUSSION TITLE
Initializing global things on server start up

### DISCUSSION THREAD

**Discussion by bcspragu** [nuxt/nuxt#16628] [Questions]
Hello!

I'm trying to configure Firebase auth on the server, it's used in server middleware to validate incoming JWTs. Prior to RC 11 the approach we used was a plugin with some global state, e.g.:

```ts
// plugins/firebase.server.ts
import { initializeApp } from 'firebase-admin/app'
import type { Auth } from 'firebase-admin/auth'
import { getAuth } from 'firebase-admin/auth'

let sauth: Auth | undefined = undefined
export default defineNuxtPlugin(({ $config }) => {
  // Plugins run on every request, but Firebase errors out if it's initialized
  // more than once, so we only initialize it once per server start.
  if (!sauth) {
    const { firebaseConfig } = $config
    const app = initializeApp(firebaseConfig)
    sauth = getAuth(app)
  }

  return {
    provide: {
      sauth,
    },
  }
})
```

And then we use `$sauth` without issue in our middleware. But recently, when I refresh a page, I see the following in my dev server:

```
[nuxt] [request error] [unhandled] [500] The default Firebase app already exists. This means you called initializeApp() more than once without providing an app name as the second argument. In most cases you only need to call initializeApp() once. But if you do want to initialize multiple apps, pass a second argument to initializeApp() to give each app a unique name.
```

And then everything is broken. I'm not sure if it was RC11 in particular that broke things, or general refactoring of this code (which was needed to make Firebase work with RC11), but the overall approach hasn't changed.

It seems like the correct solution would be to truly initialize Firebase once in some global place, and use that everywhere, but I haven't been able to figure out how to do that using App or Nitro hooks.

EDIT: For this particular case, it seems like relying on Firebase's internal store/cache works here, e.g.:

```ts
import { initializeApp, getApp } from 'firebase-admin/app'
import { getAuth } from 'firebase-admin/auth'

export default defineNuxtPlugin(({ $config }) => {
  // Plugins run on every request, but Firebase errors out if it's initialized
  // more than once, so we only initialize it once per server start.
  try {
    getApp()
  } catch {
    initializeApp($config.firebaseConfig)
  }

  return {
    provide: {
      sauth: getAuth(),
    },
  }
})
```

**Comment by justwiebe** [comment]
Did you figure this out? I'm unable to find a hook that will run when the nitro server is started up with `nuxt preview`
